crates/bevy_debug_overlays/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+        if !wgpu_features.contains(WgpuFeatures::TIMESTAMP_QUERY) {
+            panic!("DebugOverlaysPlugin added but RenderPlugin::wgpu_settings did not contain WgpuFeatures::TIMESTAMP_QUERY.");
+        }
+
+        if env::var("GPU_LOCKED").unwrap_or_default() != "magic" {
+            panic!("GPU clocks were not locked. Unable to get stable profiling results. Please rerun via the locking script.")
+        }
+    }
+}
+
+impl Default for DebugOverlaysPlugin {
+    fn default() -> Self {
+        Self {
+            ui_update_frequency: Duration::from_millis(100),
+        }
+    }
+}
+
+#[derive(Component)]
+struct GpuTimerLabelMarker;
+
+#[derive(Component)]
+struct GpuTimerDurationMarker;
+
+fn setup_gpu_time_overlay(mut commands: Commands) {
+    commands
+        .spawn(NodeBundle {
+            style: Style {
+                width: Val::Percent(100.0),
+                align_items: AlignItems::Start,
+                ..default()
+            },
+            ..default()
+        })
+        .with_children(|parent| {
+            parent
+                .spawn(NodeBundle {
+                    background_color: Color::rgba(0.10, 0.10, 0.10, 0.8).into(),
+                    style: Style {
+                        padding: UiRect::all(Val::Px(8.0)),
+                        flex_direction: FlexDirection::Column,
+                        ..default()
+                    },
+                    ..default()
+                })
+                .with_children(|parent| {
+                    parent.spawn(TextBundle::from_sections([
+                        TextSection::new(
+                            "GPU Time",
+                            TextStyle {
+                                font_size: 16.0,
+                                color: Color::WHITE,
+                                ..default()
+                            },
+                        ),
+                        TextSection::new(
+                            " (ms)  Mean     Min      Max\n",
+                            TextStyle {
+                                font_size: 10.0,
+                                color: Color::WHITE,
+                                ..default()
+                            },
+                        ),
+                    ]));
+                    parent.spawn(NodeBundle::default()).with_children(|parent| {
+                        let style = TextStyle {
+                            font_size: 12.0,
+                            color: Color::WHITE,
+                            ..default()
+                        };
+                        parent.spawn((
+                            GpuTimerLabelMarker,
+                            TextBundle::from_section("", style.clone()),
+                        ));
+                        parent.spawn((GpuTimerDurationMarker, TextBundle::from_section("", style)));
+                    });
+                });
+        });
+}
+
+fn draw_gpu_time_overlay(
+    gpu_timers: Res<GpuTimerScopes>,
```

Comment:
For the UI part specifically, bevy_render depends on on bevy_diagnostics, but bevy_ui depends on bevy_render, so it would be circular dependencies :/.

There's no reason the aggregating statistics part couldn't be on bevy_diagnostics though. We could do something like bevy_diagnostics aggregates the data produced by bevy_render, and then in another crate (a new one? bevy_core_pipeline?) we do the visual UI.

Reply:
Yeah the visualization logic definitely shouldn't be in bevy_diagnostics, bevy_render, or bevy_ui.

This should be the dep tree:
`bevy_render -> bevy_diagnostics`
`nice_diagnostic_visualizer -> bevy_diagnostics`
`nice_diagnostic_visualizer -> bevy_ui -> bevy_render`

I think we'll probably also want a "diagnostic category label" system so we can dump a bunch of render pass diagnostics in some "gpu timing" category. Then you could do something like `DiagnosticVisualizerPlugin::new().show(GpuTimingDiagnostics)`

Reply:
Not necessarily saying that this needs to be done in this PR, but it would make all of this visualization work generically useful / doesn't introduce quality debt. I can also take a stab if you want.